#95 Multi-order Submission

bh2smith opened the pull request:
Closes #95

Test Plan - Unit tests.

bh2smith committed:
First version of multi-sell order

bh2smith committed:
First version of multi-sell order

bh2smith committed:
place Multi-sell order with generic passing and generic failing unit test

bh2smith committed:
place Multi-sell order with generic passing and generic failing unit test

josojo reviewed (approved):
I like the PR because it is very readable and reuses a lot of existing code. 
However, it could be written more gas efficiently. Just mentioning it here.
 I guess readability is currently more important, gas improvements can still be done later.

bh2smith commented:
@josojo 

> However, it could be written more gas efficiently

Do you have a suggestion how it could be more gas efficient? Notice that all the `require` statements should be applied to each order (since they could have a wrong `tokenId` in any one of the orders. I have considered your suggestion, but didn't see too many places when gains could be made.

josojo commented:
At least the following three steps could be omitted:
1. onlyRegisteredAccounts() modifier
2. getting the accountId, its always the same
3. Throwing only one log statement with arrays instead of several ones.
4. Not sure, whether the compiler is smart enough to not store the hash update with a sstore, but just keeping it in memory.

But still, I do not think that it is worth the effort.

bh2smith commented:
@josojo  I completely agree with the `accoutId` retrieval, but don't necessarily like the single event (because we would then have to propagate all these changes to the listener as well). Actually this could be a really good refactoring improvement to give to a new employee during "onboarding". I am not sure what is the problem with `onlyRegistered`. You mean the fact that it is being called for every single order instead of just once? Yes, I agree this is also less efficient and should probably be changed. 

Summary, 

1. Should probably be changed. Good task for new contributor?
2. Should also be changed (change could be made in the same place as Item 1)
3. Would require more changes in the Listener as well [Really good onboarding task for new contributor] Should we create an issue about this  and label it "good first issue"?

bh2smith commented:
Just waiting on @fleupold

fleupold commented:
What was our consensus on using packed blobs instead of multiple arrays which have to implicitly aligned? 
I believe in #97 @josojo offered to change this immediately. Maybe we can include it here so given that the PR is quite small.

Otherwise I agree with 1), 2), and 4). Making this more gas efficient could be a good "ToDo" task. I do prefer reusing the same event as explained in #101

bh2smith commented:
I was under the impression that Blob order data would be changed after these draft PRs were accepted. I don't consider it such a "small change" and I didn't cant see any part of #97 regarding this immediate change.

fleupold commented:
Sure, however looking at the current diff most of the code written here (especially the unit tests) will be affected by that change. So not sure if it's worth committing this before. But if you want to feel free.

Yet, I would consider this task part of the original issue.

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

bh2smith committed:
Introduce second solidity linter (Solium - which has stronger checks for trailing whitespace, and block members.

bh2smith committed:
run solium linter on travis build

bh2smith committed:
update dev dependencies

bh2smith committed:
other annoying errors (like no empty blocks)

bh2smith committed:
Merge branch 'whitespace2' into 95_multiorder_submission

bh2smith committed:
Include new contract Library (BytesLib)

bh2smith committed:
can't access registeredAccounts from SnappAuction

bh2smith committed:
multi-order packed Encoding.

bh2smith committed:
Merge branch '95_multiorder_submission' of github.com:gnosis/dex-contracts into 95_multiorder_submission

# Conflicts:
#	contracts/SnappAuction.sol

fleupold commented on .solhintignore:
```diff
@@ -1,5 +1,7 @@
 contracts/Migrations.sol
 .solhint.json
 .solhintignore
+.soliumignore
```
unrelated changes. Can you base this diff off the one that introduced those to make reviewing easier?

bh2smith replied:
This branch assumes whitespace2 has been merged. I don't plan on reversing these.

fleupold replied:
Sure, in the future you could base your PR on the one that it assumes will have been merged before. Otherwise it's really hard for your reviewers to see which changes are relevant to this PR.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -265,4 +268,46 @@ contract SnappAuction is SnappBase {
+    }
+
+    function unpackAndVerifyOrderData(bytes memory packedOrder) private view {
+        // TODO - use ECRecover for signatures from first 16 bits of packedOrder
```
When using ECRecover we need to attach a signature to each order, which is 65 bytes of extra data. This will allow us to recover the ethereum public address which we can then use to look up the accountId. Check https://github.com/protofire/zeppelin-solidity/blob/master/contracts/ECRecovery.sol for more info on how ecrecover works.

bh2smith replied:
Thanks, although I was aware of how ECRecover works. I suppose the 16 bits is what threw you off.

fleupold replied:
Yes, the comment is weird. It sounded as if you want to recover something from 16 bits, where really you need 65 bytes of data and can only recover the main Ethereum address which still needs to be looked up. Can you adjust the comment?

bh2smith replied:
```suggestion
        // TODO - use ECRecover from signature contained in first 65 bytes of packedOrder
```

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -265,4 +268,46 @@ contract SnappAuction is SnappBase {
+        // TODO - use ECRecover for signatures from first 16 bits of packedOrder
+        uint16 accountId = publicKeyToAccountMap(msg.sender);
+        // uint16 accountId = BytesLib.toUint16(orderData, 0);
+        // WARNING - currently this enables any registered member to place an order for any other registered member!
```
For now, I'd just make sure it's msg.sender until we introduce proper signature checking.

bh2smith replied:
Do you mean require that we should fetch the accountId of msg.sender and ensure that they are registered? Meaning use the modifier `onlyRegistered()`

bh2smith replied:
Can you tell me how you are able to comment on multiple lines!

fleupold replied:
> Can you tell me how you are able to comment on multiple lines!

In solidity or in github? I don't think I can comment on multiple lines in github. In solidity using `/* ... */`

> Do you mean require that we should fetch the accountId of msg.sender and ensure ...

yes, and ensure that accountId is that value. In fact looking at the code this is what is happening becuase the other line is commented out. Is the comment still relevant?

bh2smith replied:
Im trying to see how you are able to comment on 4 lines of code here at the once?

<img width="684" alt="Screenshot 2019-06-08 at 21 40 55" src="https://user-images.githubusercontent.com/11778116/59151613-401cb100-8a36-11e9-8b63-7143953cea07.png">

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -265,4 +268,46 @@ contract SnappAuction is SnappBase {
+        uint8 sellToken = BytesLib.toUint8(packedOrder, 2 + 1);
+
+        // Would have preferred to use uint 96 or 128, but BytesLib doesn't have this.
+        uint buyAmount = BytesLib.toUint(packedOrder, 2 + 1 + 1);
```
I think you can still explicitly cast down to uint96

bh2smith replied:
I was not able to cast down to 128, I don't see how I will be able to cast down even more.

fleupold replied:
Does this not work:

```
uint foo = 1;
uint96 bar = uint96(foo);
```

?

bh2smith replied:
Unfortunately not exactly. I can confirm that 

```
uint _c = 2;
uint24 c = uint24(_c);
```
Does work, but that
```
uint _c = BytesLib.toUint(_data, 1 + 2);
uint24 c = uint24(_c);
```

does not. Actually, now I am beginning to wonder if this is an issue with `BytesLib.toUint`

bh2smith replied:
Yes, it turns out, that the bytesLib `toUint` function has a length requirement on the input bytes length as seen here

https://github.com/GNSPS/solidity-bytes-utils/blob/14ca2bd38f96f3b18c06c0cf13df1997a3831a78/contracts/BytesLib.sol#L337-L346

This means I will have to slice and pad the info and then cast down. Luckily, this will allow us to work with the desired 96 bits right off the bat, and no need to fiddle around with 128 checks.

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -179,20 +182,20 @@ contract SnappAuction is SnappBase {
-        require(numOrders == sellAmounts.length, "numOrders != sellAmounts length");
+    function placeMultiSellOrder(bytes memory packedOrders) public {
+        // Note that this could result failure of all orders if even one fails.
+        require(packedOrders.length % 27 == 0, "Each order should be packed in 27 bytes!");
```
Why 27? Isn't it 2 for accountId, 1 for buy + 1 for sell token & 12 for buy + 12 for sell token? That would be 28 bytes, wouldn't it?

bh2smith replied:
Yes it should be 28... Not sure where I lost one.

bh2smith committed:
revert import statement changes.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -179,20 +182,20 @@ contract SnappAuction is SnappBase {
+            bytes32 nextAuctionHash = sha256(abi.encodePacked(auctions[auctionIndex].shaHash, orderData));
+            auctions[auctionIndex].shaHash = nextAuctionHash;
+
+            emit PackedSellOrder(auctionIndex, auctions[auctionIndex].size, orderData);
```
Can we for now just unpack the order and call the original `postSellOrder` function n times? Otherwise let's refactor postSellOrder to call into `placeMultiSellOrder`. At the end I'd really just like to maintain one unreserved sell order event in the listener.

bh2smith replied:
No, because I can't get `uint128` from the unpacking function.

josojo replied:
Yeah, it is easier to refactor `postSellOrder`, agreed.

bh2smith replied:
Cool, but we have a gas issue to deal with first.
I have concluded that either 
- commenting `unpackAndVerifyOrderData(orderData);` or
- replacing `bytes32 nextAuctionHash = sha256(abi.encodePacked(auctions[auctionIndex].shaHash, orderData));` with `bytes32 nextAuctionHash = 0x0`

would both work...

```js
function placeMultiSellOrder(bytes memory packedOrders) public onlyRegistered() {
    // Note that this could result failure of all orders if even one fails.
    require(packedOrders.length % 28 == 0, "Each order should be packed in 28 bytes!");

    for (uint i = 0; i < packedOrders.length / 28; i++) {
        bytes memory orderData = packedOrders.slice(28*i, 28*(i+1));
        unpackAndVerifyOrderData(orderData);
        orderBatchUpdate();

        bytes32 nextAuctionHash = sha256(abi.encodePacked(auctions[auctionIndex].shaHash, orderData));
        auctions[auctionIndex].shaHash = nextAuctionHash;

        emit PackedSellOrder(auctionIndex, auctions[auctionIndex].size, orderData);
        // Only increment size after event (so it is emitted as an index)
        auctions[auctionIndex].size++;
    }
}

```

fleupold replied:
So is this related to the number of orders you are trying to place? As in does it work if you call `placeMultiSellOrder` with 0 bytes? At how many orders does it start failing? 

I guess we have to do some debugging to identify what is costing how much gas. You can use `it.only` and `truffle test -- gas` to effectively narrow down which call is causing how much gas.

bh2smith replied:
> So is this related to the number of orders you are trying to place? As in does it work if you call placeMultiSellOrder with 0 bytes? At how many orders does it start failing?

Unfortunately, the payload is not the problem here. This is a problem with contract size during migration and I haven't even started trying to unit test this with real orders yet.

We are already very well aware of the gas consumption of this contract on deployment. The following is the gas cost for deployment of the current version of the contract

```
   Deploying 'SnappAuction'
   ------------------------
   > transaction hash:    0x0026525ad588bd74b56e63ef6a43101f24f2b962bd1023221d2c9ad9a56d6f60
   > Blocks: 0            Seconds: 0
   > contract address:    0x9A2400BDeE35f01Deb6d6Cbdd24f0d2eAF9E92D5
   > account:             0xa0c455c7e343f2799E26f085bA8Ee671DDd2434A
   > balance:             99.69985876
   > gas used:            7010022
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.14020044 ETH
```

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

# Conflicts:
#	.soliumrc.json
#	contracts/SnappBase.sol

bh2smith committed:
using 28 = 2 + 1 + 1 + 12 + 12 bytes per order.

bh2smith committed:
temp fix contract size on deployment

fleupold commented:
I believe that the out of gas error might come from the fact that our smart contract byte code is exceeding 45kb (https://ethereum.stackexchange.com/questions/7436/large-contract-bytecode-45k-causes-out-of-gas-on-deploy) now.

This would explain why making the source code somewhat shorter does not incur the error anymore.

You can check the length of the `bytecode` field in `build/SnappAuction.json` after callign `truffle compile`

bh2smith committed:
Update contracts/SnappAuction.sol

bh2smith commented:
> I believe that the out of gas error might come from the fact that our smart contract byte code is exceeding 45kb (https://ethereum.stackexchange.com/questions/7436/large-contract-bytecode-45k-causes-out-of-gas-on-deploy) now.

Yes, @fleupold I agree, looking forward to dealing with this one. We are saving 30% if compiled "with optimization", but this is arguably not a solution. Also, apparently the optimizer is not audited.

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

bh2smith committed:
Merge branch '95_multiorder_submission' of github.com:gnosis/dex-contracts into 95_multiorder_submission

bh2smith committed:
completely remove placeSellOrder (as it is only needed to call the multi-order).
Can only place orders with bytes now.
Unit tests updated
No contract optimization needed!

bh2smith committed:
Delete commented unused code.

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

# Conflicts:
#	contracts/SnappBase.sol

bh2smith committed:
update package lock

bh2smith committed:
solhint disable inline assembly

bh2smith committed:
unnecessary lint config

bh2smith committed:
No need for solium security/no-inline-assembly since its covered by solhint.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -248,4 +259,14 @@ contract SnappAuction is SnappBase {
         });
     }
+
+    function orderBatchUpdate() private {
```
`createNewPendingBatchIfNecessary`? `orderBatchUpdate` sounds a bit vague and it's not really clear from the name what the difference between this method and `createNewPendingBatch` is.

bh2smith replied:
Yea cool!

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -149,34 +151,44 @@ contract SnappAuction is SnappBase {
-        }
-
-        // Update Auction Hash based on request
+    function placeSellOrders(bytes memory packedOrders) public onlyRegistered() {
```
I'd leave the old interface for convenience and just call into the new method from within the old method (creating a batch of 1 item)

bh2smith replied:
Unfortunately not. That was the reason for the large contract size that wouldn't migrate. Do you still think is should be left here?

bh2smith replied:
That is to say, if we leave the old interface, then the contract won't migrate without optimization enabled. Is this what we really want?

fleupold replied:
I see. I believe we should mitigate the large contract size for now by enabling optimisation (or even increasing the block gas limit on ganache), since otherwise we will reach the limit in the next PR again.

I will make it my priority to refactor the SnappBase into a library to give us a bit more breathing space long term for deploying within the gas limit.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -149,34 +151,44 @@ contract SnappAuction is SnappBase {
+                    encodeOrder(accountId, buyToken, sellToken, buyAmount, sellAmount)
+                )
+            );
+            auctions[auctionIndex].shaHash = nextAuctionHash;
```
Not necessarily needed for now, but we should probably note that we could make this more fas efficient by only writing to the struct when we actually need to (upon new batch creation or in the last round).

bh2smith replied:
Sorry, I don't understand what is "not needed here". Are you saying that we don't need to update the `orderBatch` hash here?

fleupold replied:
Writing `nextAuctionHash` to storage (which costs 5k gas) only needs to happen in the very last iteration of the loop (or if we create a new pending batch).

Right now it's happening once per loop.

However, making that change is not that trivial, so I'd recommend just adding a TODO for now.

Commit:
only emit SellOrder once! Whoops

josojo commented on contracts/SnappAuction.sol:
```diff
@@ -149,34 +151,41 @@ contract SnappAuction is SnappBase {
+            }
+            orderBatchUpdate();  // Could this be done more efficiently?
+            bytes32 nextAuctionHash = sha256(
+                abi.encodePacked(
```
Why do we use an encodePacked instead of encode? Is there a risk if a sha hash starts with a zero?

bh2smith replied:
I don't understand your question. We have always been using `encodePacked` and I was just continuing to do so.

josojo replied:
Yeah, we have always been using encodePacked, 
I was wondering, whether there is a risk in case we are getting leading zeros in the sha hash and in the id. In this case the backend infrastructure would not be aware of where to truncate. I think it could cause problems and we might want to switch to encode only.
However, it might all work fine, as every string starts with 0x anyways and then there will not be any truncating.  But then we can use encode anyways, instead of encodePacked

bh2smith replied:
Perhaps, this would be better in the design discussion. I would love to learn more about this and see some examples/potential points of failure.

josojo commented:
Very good progress. 
Another test which is calculating the sha hash offline and compares it to the sha hash in solidity would give us more confidence that the assembly code is working. ( not sure, how much work that is and whether it is really needed)

bh2smith commented:
@josojo I am confident that the assembly code is working since its essentially a generalized version of this (e.g. `toUint32`)

https://github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol

Furthermore, I believe that this can be tested by checking the value of the amounts in the events that are emitted in a unit test.

bh2smith committed:
more intuitive naming

fleupold commented on test/snapp_auction.js:
```diff
@@ -118,14 +103,32 @@ contract("SnappAuction", async (accounts) => {
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+      const order1 = encodeOrder(0, 1, 3, 4)
+      const order2 = encodeOrder(1, 0, 1,   123456789)
+      // const order2 = encodeOrder(1, 0, 1,   100000000000000009) --> shows 100000000000000020
```
Do we really want to commit this?

bh2smith replied:
I guess not, but I really want to make it clear that this is a pending topic of concern. I will create an issue for this instead.

fleupold reviewed (approved):
Looks good to me. I'd put back the convenience interface and enable optimisations for now. I will try to fix the gas issue asap.

bh2smith committed:
Use contract optimization (temporary fix)

bh2smith committed:
remove unused comments in Generic Multo-order test.

bh2smith committed:
Include TODO for non-trivial auction.shaHash update happening on each iteration of the loop (not necessary).

bh2smith committed:
replace original placeSellOrder interface

bh2smith committed:
Merge branch 'master' into 95_multiorder_submission

# Conflicts:
#	test/snapp_auction.js

bh2smith committed:
2^100 checks are no longer necessary!

bh2smith committed:
use utility function.

fleupold commented on contracts/SnappAuction.sol:
```diff
@@ -151,6 +151,29 @@ contract SnappAuction is SnappBase {
         emit StandingSellOrderBatch(currentBatchIndex, accountId, buyTokens, sellTokens, buyAmounts, sellAmounts);
     }
 
+    function placeSellOrder(
```
I was hoping this function could just call the new function with a byte array containing the single order encoded (instead of duplicating the logic).

bh2smith replied:
Yes, that would have been nice. Too bad this wasn't mentioned. What do you suggest I do now?

fleupold replied:
I tried to mention it in https://github.com/gnosis/dex-contracts/pull/100#discussion_r292090784

> I'd leave the old interface for convenience and just call into the new method from within the old method (creating a batch of 1 item)

We can create a new PR with just that small improvement.

bh2smith replied:
Yes very nice!

bh2smith replied:
I would argue that this isn't going to work very well with the double application of `onlyRegistered` and the fact that we have to write an entirely different version of "encodeOrder" that make 26 bytes out of the order data.

From my side, this function would be better off dead.

fleupold replied:
I'll take a crack at it if you don't mind.

bh2smith replied:
Although this is still far from correct, this is the best I can do in terms of this suggestion
```
    function placeSellOrder(
        uint8 buyToken,
        uint8 sellToken,
        uint96 buyAmount,
        uint96 sellAmount
    ) public onlyRegistered() {
        uint orderInt = uint(buyToken) + (uint(sellToken) << 8) + (uint(sellAmount) << 16) + (uint(buyAmount) << 112);
        bytes memory orderBytes = new bytes(26);
        assembly { mstore(add(orderBytes, 26), orderInt) }
        placeSellOrders(orderBytes);
```

bh2smith replied:
Yes, please @fleupold
